bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala

```diff
@@ -3,57 +3,87 @@ package uk.ac.wellcome.platform.storage.bagauditor.versioning
 import java.time.Instant
 import java.util.UUID
 
-import cats.implicits._
+import cats.{Id, Monad, MonadError}
 import uk.ac.wellcome.platform.archive.common.bagit.models.ExternalIdentifier
 import uk.ac.wellcome.platform.archive.common.ingests.models.{
   CreateIngestType,
   IngestID,
   IngestType,
   UpdateIngestType
 }
-import uk.ac.wellcome.platform.archive.common.versioning.IngestVersionManager
+import uk.ac.wellcome.platform.archive.common.versioning.{
+  IngestVersionManager,
+  IngestVersionManagerError
+}
+import uk.ac.wellcome.platform.storage.bagauditor.models._
 import uk.ac.wellcome.storage.{FailedProcess, LockDao, LockingService}
 
-import scala.util.Try
-
-class IllegalVersionAssignment(message: String)
-    extends RuntimeException(message)
-
 class VersionPicker(
-  lockingService: LockingService[Int, Try, LockDao[String, UUID]],
+  lockingService: LockingService[Either[IngestVersionManagerError, Int],
+                                 Id,
+                                 LockDao[String, UUID]],
   ingestVersionManager: IngestVersionManager
 ) {
   def chooseVersion(
     externalIdentifier: ExternalIdentifier,
     ingestId: IngestID,
-    ingestType: IngestType = CreateIngestType,
+    ingestType: IngestType,
     ingestDate: Instant
-  ): Try[Int] =
-    lockingService
+  ): Either[VersionPickerError, Int] = {
+    val assignedVersion: Id[lockingService.Process] = lockingService
       .withLocks(Set(s"ingest:$ingestId", s"external:$externalIdentifier")) {
         ingestVersionManager.assignVersion(
           externalIdentifier = externalIdentifier,
           ingestId = ingestId,
           ingestDate = ingestDate
         )
       }
-      .map {
-        case Right(version) =>
-          checkVersionIsAllowed(ingestType, assignedVersion = version)
-          version
 
-        case Left(FailedProcess(_, err)) => throw err
-        case Left(err)                   => throw new RuntimeException(s"Locking error: $err")
-      }
+    // This is a double Either: the outer Either defines the result of the locking
```

Comment:
https://typelevel.org/cats/datatypes/nested.html

Might come in handy next time!

Reply:
Next time!